CAIPs/caip-294.md

```diff
@@ -0,0 +1,431 @@
+Additionally this aligns the messaging for browser wallets to leverage existing standards for handshake (CAIP-25) and signing (CAIP-27). Thus introducing a solution focused on optimizing interoperability for multiple Wallet Providers, fostering fairer competition by reducing the barriers to entry for new Wallet Providers, and enhancing user experience across all blockchain networks.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC-2119].
+
+### Definitions
+
+Wallet Provider: A user agent that manages accounts and facilitates transactions with a blockchain.
+
+Decentralized Application (dapp): A web page that relies upon one or many Web3 platform APIs which are exposed to the web page via the Wallet.
+
+Blockchain Library: A library or piece of software that assists a dapp to interact with a blockchain and interface with the Wallet.
+
+### Messaging APIs
+
+Browser Extensions make use of events with `window.dispatchEvent` and `window.addEventListener` which will be necessary for publishing and listening messages, respectively, between the Blockchain Library used by the Decentralized Application to communicate with the Wallet Provider.
+
+This provides the foundation for any Wallet Provider to interface with a Decentralized Application using a Blockchain Library which implements this standard.
+
+Different loading times can be affected by multiple factors, which makes it non-deterministic to publish and listen to messages from different sources within the browser.
+
+#### Discovery
+
+Both Wallet Providers and blockchain libraries must listen to incoming messages that might be published after their initialization. Additionally both Wallet Providers and blockchain libraries must publish a message to both announce themselves and their intent to connect, respectively.
+
+Here is the expected logic from the Blockchain Library:
+
+```typescript
+interface WalletMapEntry {
+  params: WalletAnnounceRequestParams;
+  eventName: string;
+}
+
+const wallets: Record<string, WalletMapEntry> = {}
+
+// Blockchain Library starts listening on init
+window.addEventListener("caip294:wallet_announce", (event) => {
+  // when an announce message was received then the library can index it by uuid
+  wallets[event.detail.params.uuid] = {
+    params: event.detail.params,
+    eventName: event.detail.params.uuid
+  }
+});
+
+// Blockchain Library publishes on init
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_prompt",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_prompt",
+      params: {
+        // optionally the Blockchain Library can prompt wallets to announce matching only the chains
+        chains: []  // optional
+        //  if the Blockchain Library supports CAIP-275 then it can include a name
+        authName: "", // optional
+      },
+    }
+  }
+));
```

Comment:
Just a note [I have a related CAIP](https://github.com/ChainAgnostic/CAIPs/pull/308) that is specifically handling the case of discovery when the wallet's API is delivered via `externally_connectable`. Since in this case[ the dapp will need the `extensionId`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connect#extensionid) in order to establish a channel of communication with the wallet, the `extensionId` is a critical piece of data for us to announce in this event. Perhaps it makes sense to merge this in as an optional param to this CAIP with some of the language from my CAIP to explain its use.

Reply:
Should this new proposal be labelled as CAIP-296 which uses externally_connectable as a transport for CAIP-282 discovery??

https://github.com/ChainAgnostic/CAIPs/issues/296

Reply:
merged/addressed here: https://github.com/ChainAgnostic/CAIPs/pull/282/commits/bc4942857a8e04593ed92f7dc66653577a1c4435